lib/jsdom/browser/new-resource-loader.js

```diff
@@ -0,0 +1,131 @@
+"use strict";
+
+const fs = require("fs");
+const { parseURL } = require("whatwg-url");
+const { fromURLRecord } = require("data-urls");
+const request = require("request-promise-native");
+const { version: packageVersion } = require("../../../package.json");
+const DEFAULT_USER_AGENT = `Mozilla/5.0 (${process.platform}) AppleWebKit/537.36 (KHTML, like Gecko) ` +
+  `jsdom/${packageVersion}`;
+const IS_BROWSER = Object.prototype.toString.call(process) !== "[object process]";
+
+class ResourceLoader {
+  constructor() {
+    this._defaultOptions = {
+      agentOptions: {
+        keepAlive: true,
+        keepAliveMsecs: 115 * 1000
+      },
+      pool: { maxSockets: 6 },
+      strictSSL: true
+    };
+  }
+
+  _readFile(filePath) {
+    let readableStream;
+    let abort; // Native Promises doesn't have an "abort" method.
+
+    /*
+     * Creating a promise for two reason:
+     *   1. fetch always return a promise.
+     *   2. We need to add an abort handler.
+    */
+    const promise = new Promise((resolve, reject) => {
+      readableStream = fs.createReadStream(filePath);
+      let data = Buffer.alloc(0);
+
+      abort = reject;
+
+      readableStream.on("error", reject);
+
+      readableStream.on("data", chunk => {
+        data = Buffer.concat([data, chunk]);
+      });
+
+      readableStream.on("end", () => {
+        resolve(data);
+      });
+    });
+
+    promise.abort = () => {
+      readableStream.destroy();
+      const error = new Error("request canceled by user");
+      error.isAbortError = true;
+      abort(error);
+    };
+
+    return promise;
+  }
+
+  /**
+   * Returns the options to make the request
+   * @param {object} options - Custom options for the request.
+   */
+  _getRequestOptions(options = {}) {
+    const rqOptions = options.requestOptions;
+    const requestOptions = {
+      agent: rqOptions.agent,
+      agentOptions: rqOptions.agentOptions ?
+        Object.assign({}, this._defaultOptions.agentOptions, rqOptions.agentOptions) :
```

Comment:
You can just always do the `Object.assign`; it'll work the same way. Here and below.